docs/TRACEZ_DESIGN.md

```diff
@@ -0,0 +1,175 @@
+
+## Proposed Design
+
+We have divided the implementation of the /tracez and /traceconfigz zPages into two frontend-centric
+and backend-centric components.
+
+### Frontend
+
+#### HttpHandler
+
+The `HttpHandler` is responsible for rendering corresponding HTML content. An abstract base class,
+`ZPageHandler` (OpenCensus implementation), is implemented to standardize handlers for different
+zPages. Each page will implement their own `ZPageHandler`, extending the base class, to generate the
+corresponding HTML content for that page.
+
+![httphandler](img/httphandler.png)
+
+##### TraceZ Handler
+
+For the `TracezZPageHandler` class, the span data from `TracezDataAggregator` will be passed in when
+an instance of the class is created. It will later be used to retrieve span information and display
+the data in a table.
+
+##### TraceConfigZ Handler
+
+For the `TraceConfigzZPageHandler` class, the `TraceConfig` class will be used to change sampling
+rate and tracing parameters.
+
+#### HttpServer
+The `HttpServer` is responsible for listening to incoming requests, obtaining requested data, and
+rendering corresponding HTML content. The `HttpServer` class from `com.sun.net` will be used to
+handle http requests and responses on different routes (users need to ensure that they are using a
+version of JDK that includes the `HttpServer` class; this requirement will be added to the README).
+Once a request is received by the `HttpServer`, it will invoke the handle function which in turn
+invokes the `emitHtml` function to render the HTML content.
+
+The `HttpServer` class utilizes `com.sun.net.httpserver` to create servers. Users need to ensure
+that they are using a version of JDK that comes with the package.
+
+##### Handling Requests
+
+![requests-flowchart](img/requests-flowchart.png)
+
+### Backend
+
+#### Overview
+
+The proposed structure is encompassed by the following diagram:
+
+![span-lifecycle](img/span-lifecycle.png)
+
+Spans, which are the units of tracing, are monitored by a `SpanProcessor`. The `SpanProcessor` is
+exposed to a `DataAggregator`, which can retrieve information about the spans through API calls.
+Lastly, the frontend calls functions in the `DataAggregator` to obtain information needed for the
+web page.
+
+#### SpanProcessor
+
+A `SpanProcessor` watches the lifecycle of each span, and its functions are called every time a span
+starts or ends. For the /tracez zPage, I have implemented a `TracezSpanProcessor`, which will
+maintain two data structures: a running span cache and a completed span cache.
+
+##### Visual Diagram
+
+Below is a visual diagram of the `TracezSpanProcessor` class:
+
+![span-processor-flowchart](img/span-processor-flowchart.png)
+
+When a span starts, it is first added to the `runningSpanCache`. Once that span ends, it is removed
+from the `runningSpanCache` and added to the `completedSpanCache` as either a latency sample if
+there are no errors or an error sample if there are.
+
+##### Initial Design
```

Comment:
The information under the "proposed design" sections I think would make for good OSS blog post material.

Reply:
After reading the other OSS blog posts, I'm a little skeptical of whether we should include implementation details. Most of the posts just talk about a general overview of the product and then describe its use cases and how to start using it.

Reply:
That sounds good with me.